artifactory/commands/buildinfo/publish.go

```diff
@@ -248,3 +275,187 @@ type partialModule struct {
 	artifacts    map[string]buildinfo.Artifact
 	dependencies map[string]buildinfo.Dependency
 }
+
+// Set build property for each dependency.
+func addDependenciesProps(partialBuildsInfo []*buildinfo.PartialBuildInfo, sm *artifactory.ArtifactoryServicesManager, threads int) error {
+	// Key: dependency sha1, Value: pointer to dependency build property.
+	depBuildMap := make(map[string][]*string)
+	var searchResults []servicesutils.ResultItem
+	// Search dependencies for all partials build info.
+	for _, partialBuildInfo := range partialBuildsInfo {
+		repositoriesSet, err := getLocalRepos(partialBuildInfo.ResolverRepositories, sm)
+		if len(repositoriesSet) == 0 {
+			if err != nil {
+				log.Warn("Fail to find repositories to collect dependencies build-name & build-number", err)
+			}
+			continue
+		}
+		sha1Set := commandutils.NewStringSet()
+		// Map the dependencies sha1 to their Build property.
+		for _, module := range partialBuildInfo.Modules {
+			for i := 0; i < len(module.Dependencies); i++ {
+				if module.Dependencies[i].Checksum != nil {
+					// Sha1 may be present in more than a single module.
+					depBuildMap[module.Dependencies[i].Sha1] = append(depBuildMap[module.Dependencies[i].Sha1], &module.Dependencies[i].Build)
+					sha1Set.Add(module.Dependencies[i].Sha1)
+				}
+			}
+		}
+		partialSearchResults, err := getArtifactsPropsBySha1(repositoriesSet, sha1Set, sm, threads)
+		if err != nil {
+			log.Warn("Fail to search artifacts by sha1 from Artifactory", err)
+			continue
+		}
+		searchResults = append(searchResults, partialSearchResults...)
+	}
+	// Update dependencies properties according to sha1.
+	for _, searchResult := range searchResults {
+		var buildName, buildNumber, timestamp string
+		for _, prop := range searchResult.Properties {
+			switch prop.Key {
+			case "build.name":
+				buildName = prop.Value + "/"
+			case "build.number":
+				buildNumber = prop.Value + "/"
+			case "build.timestamp":
+				timestamp = prop.Value
+			}
+		}
+		for _, prt := range depBuildMap[searchResult.Actual_Sha1] {
+			*prt = buildName + buildNumber + timestamp
+		}
+	}
+	return nil
+}
+
+// Search artifacts properties by sha1 among all the repositories.
+// We split the repositories and sha1s artifacts into small groups in order to send small AQls to Artifactory.
+// If an AQL is too big, 500 will be returned.
+func getArtifactsPropsBySha1(repositories []string, sha1s *commandutils.StringSet, sm *artifactory.ArtifactoryServicesManager, semaphoreBatchesSize int) (totalResults []servicesutils.ResultItem, err error) {
```

Comment:
Let's use the exiting producer-consumer utility in this function, instead of a new mechanism here.